realm/realm-library/src/objectServer/java/io/realm/mongodb/functions/Functions.java

```diff
@@ -159,7 +181,44 @@ public T run() throws ObjectServerError {
      * @see RealmAppConfiguration#getDefaultCodecRegistry()
      */
     public <T> RealmAsyncTask callFunctionAsync(String name, List<?> args, Class<T> resultClass, RealmApp.Callback<T> callback) {
-        return callFunctionAsync(name, args, resultClass, defaultCodecRegistry, callback);
+        return callFunctionAsync(name, args, defaultCodecRegistry, defaultCodecRegistry.get(resultClass), callback);
+    }
+
+    /**
+     * Call a MongoDB Realm function asynchronously with custom result decoder.
+     * <p>
+     * This is the asynchronous equivalent of {@link #callFunction(String, List, Decoder)}.
+     *
+     * @param name Name of the Stitch function to call.
+     * @param args Arguments to the Stitch function.
+     * @param resultDecoder The decoder used to decode the result.
+     * @param callback The callback that will receive the result of the request. If the request
```

Comment:
The Callback will always return `ObjectServerError` in case of an error, but it might just wrap the underlying exception. But this description is not entirely accurate.

Reply:
I see. Should we then introduce a specific _client side encoding/decoding_ related error code and also encapsulate it as `ObjectServerError`s in the synchronous case?

As I see it the most appropriate existing error code is `INVALID_PARAMETER`, but it sort of sounds like it would overlap with actual object server errors, whereas the encoding/decoding error is a client side error potentially caused by users misconfiguration of the codec registry etc.

Reply:
That probably makes a lot of sense. The ErrorCode.java has a section for binding specific error codes, that are already wrapping other Java exceptions: https://github.com/realm/realm-java/blob/v10/realm/realm-library/src/objectServer/java/io/realm/ErrorCode.java#L41

Reply:
There is also an already existing `Type.JSON` (https://github.com/realm/realm-java/blob/v10/realm/realm-library/src/objectServer/java/io/realm/ErrorCode.java#L326)  Would it make sense to use that for extended JSON too, or is it again object server internal (can't really find any use of it)?

And can I just pick the actual error codes freely as long as it does not overlap any others with the same type (whether it ends up being Java or JSON)? Or are there other locations to investigate for overlap?

Reply:
The JSON error type is an error category exposed by `App` in C++. I don't think we should use it for our own purpose.

The only requirement for error codes is that they should be unique within each section, so for `Type.JAVA`, `1000-1002` (I think) is already taken.

Reply:
Added a couple of BSON related `Type.JAVA` error codes and now throws `ObjectServerErrror`s with the specific error codes on encoding, decoding and codec configuration errors